### PR TITLE
Add GitLab as a source for repositories

### PR DESCRIPTION
**Pull request by zillemarco**
This PR adds GitLab as a source for repositories 🥳

### CODE REVIEW

**Comment by zillemarco** [comment]
@ltickett @souldzin here we go 🙂 

@lucavallin feel free to have a look and let's start from here 💪 🙂

**Comment by ltickett** [comment]
@zillemarco we may just want to review the list of projects we included (we picked a bit random/haphazzardly)

I think GitLab is the only one which meets the 1,000 stars requirement, though me and Paul had a look and it seems GitHub is used very differently to GitLab.

On GitHub repos tend to have more stars than forks. On GitLab it's the other way around?

**Comment by zillemarco** [comment]
That's a very good point 🤔 Are you proposing to update the "filter" to count the number of forks instead of stars for repositories coming from GitLab?

I'll update the projects to only keep GitLab's project for now anyway 🙂

File: firstissue.json
Comment (souldzin) [inline firstissue.json:6]: **todo:** Either here or in a follow-up, it'd be nice to improve the selector we use for GitLab repositories.

As a developer wanting to add a GitLab repository to this project, it's hard for me to tell what projects are already included since we are looking by ID. It looks like the GraphQL API doesn't support querying by multiple project paths? In the meantime maybe we could do something like `"gitlab-org/gitlab|278964"` and split on `|` to grab the ID. This way we are still capturing the project path in the meantime.

We don't have to handle this here. Just dropping a note for this.

**Comment by souldzin** [comment]
> That's a very good point 🤔 Are you proposing to update the "filter" to count the number of forks instead of stars for repositories coming from GitLab?
> 
> I'll update the projects to only keep GitLab's project for now anyway 🙂

**suggestion:** GitLab has some satellite projects that might be nice to include. I'm thinking of [`gitaly`](https://gitlab.com/gitlab-org/gitaly/), [`gitlab-ui`](https://gitlab.com/gitlab-org/gitlab-ui), and [`gitlab-web-ide`](https://gitlab.com/gitlab-org/gitlab-web-ide).

Your call @zillemarco :smile:

File: firstissue.json
Comment (zillemarco) [reply firstissue.json:6]: Thanks for the note Paul 🙂 That's actually one thing I have discussed with Lee today and we decided to use the ID until we add the possibility to filter by multiple project paths on the GitLab side 🙂 

I like your splitting idea though 🤔 I'll give it a shot (it's just a string split 😄 )

**Comment by zillemarco** [comment]
> > That's a very good point 🤔 Are you proposing to update the "filter" to count the number of forks instead of stars for repositories coming from GitLab?
> > I'll update the projects to only keep GitLab's project for now anyway 🙂
> 
> **suggestion:** GitLab has some satellite projects that might be nice to include. I'm thinking of [`gitaly`](https://gitlab.com/gitlab-org/gitaly/), [`gitlab-ui`](https://gitlab.com/gitlab-org/gitlab-ui), and [`gitlab-web-ide`](https://gitlab.com/gitlab-org/gitlab-web-ide).

Only issue with those projects is that they would get filtered out by the number of stars 😅

**Comment by ltickett** [comment]
> Are you proposing to update the "filter" to count the number of forks instead of stars for repositories coming from GitLab?

@zillemarco - no. Just that i'm not sure if we need to apply exactly the same rules to which projects we can/can't include?

We may need a bit of direction from @lucavallin - or may be given a little "artistic license" :)

File: generate.ts
Comment (lucavallin) [inline generate.ts:458]: That, and repos that don't meet criteria anymore are removed and the file is overwritten.

File: firstissue.json
Comment (lucavallin) [reply firstissue.json]: The ID is fine for now, but make sure to create a follow-up issue to replace that with the repo name

**Comment by lucavallin** [comment]
@ltickett @zillemarco Thanks for your contributions, it's looking great so far. I added a comment or two ;)
Furthermore, I am not familiar with GitLab, but it's reasonable to have a different filter for it to account for different dynamics. Come up with something sensible, "artistic license" granted.

File: firstissue.json
Comment (zillemarco) [reply firstissue.json]: @lucavallin I applied Paul's suggestion so that once we improve the GitLab endpoint we just have to remove the ID and leave the project full path (so we don't have to manually find it from the ID) 👍 🙂

File: generate.ts
Comment (zillemarco) [reply generate.ts:458]: @lucavallin would it be alright to remove this logic?
Now that we have more sources/providers it's kinda harder to recreate the file structure, and at this point we wouldn't even have all the info needed to recreate it (e.g. we don't know where a project is coming from).
Moreover, I think it would be sufficient to "hide" a project (so don't show any issue from it) when it doesn't meet the criteria anymore. I think it'd be kinda "bad" to have to re-add the project multiple times (thinking about new projects that may not be super "active" or "open").

WDYT?

File: firstissue.json
Comment (nickveenhof) [reply firstissue.json:6]: @zillemarco Added some more repositories that I manually verified that have issues with one of the said labels, however it gives an error as fullName doesn't seem to be available for all of these.

File: firstissue.json
Comment (nickveenhof) [inline firstissue.json:10]:
```suggestion
      "quick win",
      "Seeking community contribution",
      "help wanted",
      "good first issue"
```

Help wanted & good first issue are being used by other projects on GitLab

File: firstissue.json
Comment (ltickett) [reply firstissue.json:6]: @taucher2003 are we close to having the namespace permission stuff fixed?

Essentially here we're having to query group.fullPath because namespace.fullPath doesn't work unless you're a member.

File: generate.ts
Comment (zillemarco) [reply generate.ts:458]: @lucavallin that sounds totally good to me, I'll do that 🙂

**Comment by cloudhoy** [comment]
@lucavallin Thanks for the ping! 😄

Hi everyone! Just wanted to add my two cents to this discussion.

I'm wondering if lowering the star requirement for GitLab may be a little "unfair" to developers on GitHub who want to add their repositories to [firstissue.dev](https://firstissue.dev) but are unable to meet the 1,000 star requirement.

Perhaps we should just lower the star requirements across the board instead? 🤔

**Comment by nickveenhof** [comment]
> @lucavallin Thanks for the ping! 😄
> 
> Hi everyone! Just wanted to add my two cents to this discussion.
> 
> I'm wondering if lowering the star requirement for GitLab may be a little "unfair" to developers on GitHub who want to add their repositories to [firstissue.dev](https://firstissue.dev) but are unable to meet the 1,000 star requirement.
> 
> Perhaps we should just lower the star requirements across the board instead? 🤔

I'm not opposed to even lowering or removing the requirement completely. The current process is to manually add the repository to firstissue.json anyway, so I don't see why any of those filters would even need to be in the list. Perhaps a great improvement wouldbe some documentation that you need to have at least 5 issues with any of those labels present at all times to qualify to get in the list? That seems more fair? 

Also, I believe this is a separate discussion, should we move this to a new issue?

**Comment by lucavallin** [comment]
I disagree with removing the requirements: they are definitely arbitrary and can be discussed and changed, but they are to ensure some level of quality of the repos that make it to the list. The idea of [FirstIssue](https://firstissue.dev) is not only to find a project to contribute to, but to find one that is relevant, somewhat "famous", that you'd be eager to contribute to and proud of doing so, and that you'd add on your LinkedIn / resume. There are other websites like [FirstIssue](https://firstissue.dev) that are full of (mostly) irrelevant 100-stars "twitter-ios-clone" projects and we want to avoid that ;) That obviously cuts out super-awesome repos with 890 stars, but we need to draw the line somewhere.

However, if GitLab has ~1/3 of the users of GitHub, then I think it's fair to have different requirements for it, since 350 stars on GitLab = 1050 stars on GitHub (well, naïve math, but you get the point...). I am not sure what those requirements should be though 🙂.

File: firstissue.json
Comment (Taucher2003) [reply firstissue.json:6]: @ltickett sadly not, the first step of creating the new ability that keeps the member behaviour is done, but probably the hardest part is the next one, as we need to identify and check all the points, where something like `:"read_#{object_type}"` is used in the ability checks. After that we should be good to move the namespace type authorization over to the one that works as intended with public namespaces.

Comment (zillemarco) [reply firstissue.json:6]: I added all those repos except `CalcProgrammer1/OpenRGB` to avoid issues with the group full name being blocked 🙂 

We can always add it later on 🙂

File: firstissue.json
Comment (zillemarco) [reply firstissue.json:10]: Sounds good, I added those 👍

File: generate.ts
Comment (zillemarco) [reply generate.ts:458]: @lucavallin this turned out to be quite "intricate" since we have multiple reasons why a project is not added (namely stars count and  issues count) and we have different places where we do filter those out (for GitHub is a mix of GraphQL query and manual filter, for GitLab is only manual filter).

Probably, a good way to handle this would do all the queries and then iterate over the repos coming from a source and see if it's there or not, and log those results. WDYT? Anyway, in the spirit of speeding this up I would be tempted to not block this PR only for that, we can fix it as an iteration, but up to you to decide 🙂

**Comment by zillemarco** [comment]
> However, if GitLab has ~1/3 of the users of GitHub, then I think it's fair to have different requirements for it, since 350 stars on GitLab = 1050 stars on GitHub (well, naïve math, but you get the point...). I am not sure what those requirements should be though 🙂.

I used 300 stars as a "stopping point" for projects coming from GitLab (more round number than 350, plus it let a couple more projects in 😅)

File: firstissue.json
Comment (zillemarco) [reply firstissue.json:6]: Actually, `cznic/sqlite` and `timvisee/ffsend` had the same issue 😓

**Comment by zillemarco** [comment]
@lucavallin sorry it took me this long but I now have pushed up the changes and rebased with your latest changes (quite a refactor 😅)

I think this is good for you to have another look now 👍 🙂

File: generate.ts
Comment (lucavallin) [reply generate.ts]: Yep, sounds good, feel free to do that as a follow-up (please create an issue for that so we don't forget).

Comment (zillemarco) [reply generate.ts]: @lucavallin, follow-up issue created 👍

File: firstissue.json
Comment (zillemarco) [reply firstissue.json]: Follow-up issue created 👍

**Comment by lucavallin** [comment]
Hey @zillemarco! Looks good, thanks for the changes.

1. There is a conflict in config.json, please make sure new repos are synced.
2. Do we need any kind of auth for GitLab?
3. Let me know if/when you're ready to merge this

**Comment by zillemarco** [comment]
@lucavallin I just resolved the conflict, so we're good to merge if that's ok with you as well 🙂 

And there's no need for authorization for GitLab, we're using only public APIs 👍

**Review by lucavallin — APPROVED** [review]
Please double-check deployment on Vercel before merging. 
Awesome contribution - thank you for it!

**Comment by zillemarco** [comment]
@lucavallin the vercel message is for you, right? 🤔 I can't access the deployment 🤔

**Comment by lucavallin** [comment]
@zillemarco So it seems... but it is successful!

**Comment by zillemarco** [comment]
🥳 thank you @lucavallin 😄

**Pull request merged**
